src/main/scala/org/broadinstitute/dsde/workbench/sam/api/StandardUserInfoDirectives.scala

```diff
@@ -30,61 +30,62 @@ trait StandardUserInfoDirectives extends UserInfoDirectives with LazyLogging {
   implicit val executionContext: ExecutionContext
   val identityConcentratorService: Option[IdentityConcentratorService] = None
 
-  def requireUserInfo: Directive1[UserInfo] =
-    (optionalHeaderValueByName(accessTokenHeader) &
-      optionalHeaderValueByName(googleSubjectIdHeader) &
-      optionalHeaderValueByName(expiresInHeader) &
-      optionalHeaderValueByName(emailHeader) &
-      headerValueByName(authorizationHeader) &
-      provide(identityConcentratorService)
-    ) tflatMap {
+  def requireUserInfo: Directive1[UserInfo] = handleAuthHeaders(requireUserInfoFromJwt, requireUserInfoFromOIDC)
+  def requireCreateUser: Directive1[CreateWorkbenchUser] = handleAuthHeaders(requireCreateUserInfoFromJwt, requireCreateUserInfoFromOIDC)
+
+  /**
+    * Utility function that knows how to handle the request based on whether or not JWT is present.
+    *
+    * @param fromJwt function to call when jwt is present
+    * @param fromOIDC function to call when jwt is not present
+    * @tparam T type this request returns
+    * @return
+    */
+  private def handleAuthHeaders[T](fromJwt: (JwtUserInfo, OAuth2BearerToken, IdentityConcentratorService) => Directive1[T], fromOIDC: => Directive1[T]): Directive1[T] =
+    (headerValueByName(authorizationHeader) &
+      provide(identityConcentratorService)) tflatMap {
       // order of these cases is important: if the authorization header is a valid jwt we must ignore
       // any OIDC headers otherwise we may be vulnerable to a malicious user specifying a valid JWT
       // but different user information in the OIDC headers
-      case (_, _, _, _, JwtAuthorizationHeader((jwtJson, bearerToken)), Some(icService)) =>
-        onSuccess {
-          getUserInfoFromJwt(jwtJson, bearerToken, directoryDAO, icService).unsafeToFuture()
-        }
-      case (Some(token), Some(googleSubjectId), Some(expiresIn), Some(email), _, _) =>
-        // request coming through Apache proxy
-        onSuccess {
-          getUserInfoFromOidcHeaders(directoryDAO, googleSubjectId, expiresIn, email, OAuth2BearerToken(token)).unsafeToFuture()
-        }
+      case (JwtAuthorizationHeader(Success(jwtUserInfo), bearerToken), Some(icService)) =>
```

Comment:
Maybe not a big deal, but the asymmetry here is bugging me. All of the parsing of OIDC -> T is in the fromOIDC function, but this handleAuthHeaders is doing half of the parsing of the header values and leaving the rest to fromJWT after pulling out a jwtUserInfo and a bearerToken.
Can we change the signatures of requireUserInfoFromJwt & requireCreateUserInfoFromJwt to look more like requireUserInfoFromOidc & requireCreateUserInfoFromOidc? I think it's a little different in that we want to allow the authorization header to not be present, so maybe they return some optional or maybe that's handled in this 'handleAuthHeaders' function in choosing where to route it.

Or maybe this is more natural with directives.